src/libConsensus/ConsensusLeader.cpp

```diff
@@ -63,9 +63,207 @@ bool ConsensusLeader::CheckState(Action action) {
   return true;
 }
 
+bool ConsensusLeader::CheckStateSubset(uint16_t subsetID, Action action) {
+  ConsensusSubset& subset = m_consensusSubsets.at(subsetID);
+
+  static const std::multimap<ConsensusCommon::State, Action> ACTIONS_FOR_STATE =
+      {{CHALLENGE_DONE, PROCESS_RESPONSE},
+       {FINALCHALLENGE_DONE, PROCESS_FINALRESPONSE}};
+
+  bool found = false;
+
+  for (auto pos = ACTIONS_FOR_STATE.lower_bound(subset.m_state);
+       pos != ACTIONS_FOR_STATE.upper_bound(subset.m_state); pos++) {
+    if (pos->second == action) {
+      found = true;
+      break;
+    }
+  }
+
+  if (!found) {
+    LOG_GENERAL(WARNING, "SubsetID: " << subsetID << ", Action "
+                                      << GetActionString(action)
+                                      << " not allowed in state "
+                                      << GetStateString());
```

Comment:
I think what you want to log is the `m_state` of the subset, but `GetStateString()` here will return the value of `ConsensusCommon::m_state`

Reply:
Actually it might be useful to log both the leader's state as well as the state of the subset

Reply:
Yes. I agree. Will log both overall and subset’s states